transport_plugins/socket_lib/iotile_transport_socket_lib/generic/abstract_socket_implementation.py

```diff
@@ -0,0 +1,127 @@
+"""Abstract base class for different socket transport implementations
+
+Allows the AbstractDeviceAdapter and Servers to communicate via any different
+protocols depending onsystem architect's choice
+
+The functions will contain implementation details specific to the protocol.
+For example, Unix Socket's send and receive functions will encode and send the
+size of each transmission first and call the asyncio StreamReader and Writer
+send and receive functions.
+"""
+
+import abc
+import struct
+import asyncio
+
+
+class AsyncioSocketConnection:
+    """Paired read and write commands for an active Socket Connection 
+        
+    Args:
+        reader (asyncio.StreamReader): The reader returned from the connect command or callback
+        writer (asyncio.StreamWriter): The writer returned from the connect command or callback
+        logger (logging.Logger): The logger to use (Since this is used by both the Server and Client, the
+            logger should be that of the Server or Client)
+    """
+
+    _ARCHMAGIC = "ARCH".encode()
+    _HEADERFORMAT = '4sL'
+    def __init__(self, reader, writer, logger):
+        self.reader = reader
+        self.writer = writer
+        self._logger = logger
+
+    async def send(self, encoded):
+        """Send encoded byte data. First send a header with the data's length, then the data itself
+        Return once data in the buffer has been drained 
+
+        Args:
+            encoded (bytes): Encoded data to send
+        """
+        try:
+            packed_header = struct.pack(self._HEADERFORMAT, self._ARCHMAGIC, len(encoded))
+            self.writer.write(packed_header)
+            self.writer.write(encoded)
+            await self.writer.drain()
+
+        except Exception:
```

Comment:
Yeah, I couldn't track down the specific exception that asyncio would throw on some kind of write failure. Is there some listing in Python's docs of what exceptions each library function can throw?

Reply:
The question to answer is "why are we altering the Exception flow". 

If the reason is "We need to return a specific exception, because another function relies on it" -- then doing the `raise DesiredError from ActualError` is a way to capture the full stacktrace information, while still enabling program flow dependent on the rest of the program.

If the reason is "we want the program to keep going anyways", we typically don't also `raise` inside of the handler and instead log a message or take some other logic action. (this doesn't appear to be the case here).

So we typically use it in a pattern like:

```python
def send(self, encoded):
  ...
  except Exception:  # pylint:disable=bare-except; We need to raise ConnectionError here
    raise ConnectionError from Exception
```

and you might use it like:
```python
myconn = AsyncioSocketConnection(...)

try:
  ...
  myconn.send(encoded)
  ...
except ConnectionError:
  self._logger.error("what?!")
```

```python
except Exception:  # pylint:disable=bare-except; We need to log the error and not crash
  self._logger.exception("Unhandled exception caught, we should fix it", exc_info=True)
```

Otherwise, it's healthier for the code base to not handle this exception because chances are we'd be masking a bug in the code leading up to it.